dreamer/algos/dreamer_algo.py

```diff
@@ -0,0 +1,192 @@
+import torch
+import torch.optim as optim
+from rlpyt.utils.tensor import to_onehot, infer_leading_dims, restore_leading_dims
+
+from dreamer.models.rnns import get_feat, get_dist, RSSMState
+
+from rlpyt.algos.base import RlAlgorithm
+from rlpyt.samplers.collections import Samples
+from rlpyt.utils.quick_args import save__init__args
+
+
+class Dreamer(RlAlgorithm):
+
+    def __init__(
+            self,  # Hyper-parameters
+            batch_size=50,
+            batch_length=50,
+            train_every=1000,
+            train_steps=100,
+            pretrain=100,
+            model_lr=6e-4,
+            value_lr=8e-5,
+            actor_lr=8e-5,
+            grad_clip=100.0,
+            dataset_balance=False,
+            discount=0.99,
+            discount_lambda=0.95,
+            horizon=15,
+            action_dist='tanh_normal',
+            action_init_std=5.0,
+            expl='additive_gaussian',
+            expl_amount=0.3,
+            expl_decay=0.0,
+            expl_min=0.0,
+            optim_kwargs=None,
+            free_nats=3,  # TODO: Do we actually need this??
+            kl_scale=1,
+            type=torch.float,
+    ):
+        super().__init__()
+
+        self._batch_size = batch_size
+        del batch_size  # Property.
+        save__init__args(locals())
+        self.update_counter = 0
+
+        self.optimizer = None
+        self.learning_rate = model_lr
+        self.model_weight = model_lr / self.learning_rate
+        self.value_weight = value_lr / self.learning_rate
+        self.actor_weight = actor_lr / self.learning_rate
+        self.type = type
+
+    def initialize(self, agent, n_itr, batch_spec, mid_batch_reset, examples, world_size=1, rank=0):
+        self.agent = agent
+        self.optimizer = optim.Adam(self.agent.model.parameters(), lr=self.learning_rate)
+
+    def async_initialize(self, agent, sampler_n_itr, batch_spec, mid_batch_reset, examples, world_size=1):
+        self.agent = agent
+
+    def optim_initialize(self, rank=0):
+        self.optimizer = optim.Adam(self.agent.model.parameters(), lr=self.learning_rate)
+
+    def optimize_agent(self, itr, samples=None, sampler_itr=None):
+        self.loss(samples)
+
+    def loss(self, samples: Samples):
+        """
+        Compute the loss for a batch of data.  This includes computing the model and reward losses on the given data,
+        as well as using the dynamics model to generate additional rollouts, which are used for the actor and value
+        components of the loss.
+        :param samples: rlpyt Samples object, containing a batch of data.  All vectors are [timestep, batch, vector_dim]
+        :return: FloatTensor containing the loss
+        """
+        model = self.agent.model
+
+        # Extract tensors from the Samples object
+        # They all have the batch_t dimension first, but we'll put the batch_b dimension first.
```

Comment:
Now that the RSSM model uses the batch_t dimension first, we should probably get rid of the dimension-flipping in this function.